qiskit/transpiler/passes/mapping/check_cnot_direction.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+This pass checks if the CNOTs (or any other 2Q) in the DAG follow the right
+direction with respect to thecoupling map.
+"""
+
+from qiskit.transpiler._basepasses import AnalysisPass
+from qiskit.mapper import Layout
+from qiskit.extensions.standard.swap import SwapGate
+
+
+class CheckCnotDirection(AnalysisPass):
+    """
+    checks if the CNOTs (or any other 2Q) in the DAG follow the right
+    direction with respect to thecoupling map
+    """
+
+    def __init__(self, coupling_map, initial_layout=None):
+        """
+        Checks if the 2Q gates in DAGCircuit are in the allowed direction with
+        respect to `coupling_map`.
+        Args:
+            coupling_map (CouplingMap): Directed graph represented a coupling map.
+            initial_layout (Layout): The initial layout of the DAG to analyze.
+        """
+        super().__init__()
+        self.layout = initial_layout
+        self.coupling_map = coupling_map
+
+    def run(self, dag):
+        """
+        If `dag` is mapped and the direction is correct the property
+        `is_direction_mapped` is set to True (or to False otherwise).
+
+        Args:
+            dag (DAGCircuit): DAG to map.
+        """
+        if self.layout is None:
+            if self.property_set["layout"]:
+                self.layout = self.property_set["layout"]
+            else:
+                self.layout = Layout.generate_trivial_layout(*dag.qregs.values())
+
+        self.property_set['is_direction_mapped'] = True
+
+        for gate in dag.get_2q_nodes():
```

Comment:
Will `dag.get_2q_nodes` return 2q nodes that don't require the qubits be adjacent on the coupling map? e.g. `barrier`?

Reply:
according with the code in `twoQ_nodes`: `node_data['type'] == 'op' and len(node_data['qargs']) == 2`. So no requirement. It's true that a a 2Q barrier here fails. Adding a test for this and fixing it.

Reply:
In 4c8b5520 I'm skipping gates with not in `['swap', 'cx', 'CX', 'cz', 'CZ']`. I think the need to require a direction should be in the meta info of the gate. But we dont have such a thing.

Reply:
i think we should change this to `two_q_gates`, and that would return gates only. `Barrier`, `snapshot`, etc. are not instances of `Gate`, and they are generally not interesting from a mapping point of view.

Reply:
I added the `twoQ_gates` method in 6a51149a.